File: python/2003.py
from typing import List

class Solution:
    def smallestMissingValueSubtree(self, parents: List[int], nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [1] * n
        if min(nums) > 1: return ans
        startNode = nums.index(1)
        m = [[] for _ in range(n)]
        for i in range(1, n):
            m[parents[i]].append(i)
        vis = set()
        def dfs(i: int) -> None:
            vis.add(nums[i])
            for c in m[i]:
                if nums[c] not in vis:
                    dfs(c)
        mex = 2
        while startNode >= 0:
            dfs(startNode)
            while mex in vis:
                mex += 1
            ans[startNode] = mex
            startNode = parents[startNode]
        return ans

File: go/2003.go
package main

import "slices"

func smallestMissingValueSubtree(parents []int, nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	for i := 0; i < n; i++ {
		ans[i] = 1
	}
	startNode := slices.Index(nums, 1)
	if startNode < 0 {
		return ans
	}
	g := make([][]int, n)
	for i := 1; i < n; i++ {
		pi := parents[i]
		g[pi] = append(g[pi], i)
	}
	vis := make(map[int]bool, n)
	var dfs func(int)
	dfs = func(i int) {
		vis[nums[i]] = true
		for _, c := range g[i] {
			if !vis[nums[c]] {
				dfs(c)
			}
		}
	}
	mex := 2
	for startNode >= 0 {
		dfs(startNode)
		for vis[mex] {
			mex++
		}
		ans[startNode] = mex
		startNode = parents[startNode]
	}
	return ans
}
